### PR TITLE
Additional Backend Refactoring : Versioned Blocks

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
This started as a smaller PR that has gotten fairly intense.  List of changes
- Refactor of backend.Reader and Writer interfaces to be agnostic to the underlying data layout
  - Changes to all backends to match the new interfaces
- Removal of index functionality in the cli (not worth further complication of the BackendBlock interface)
- Minor refactor/simplification of CompactorBlock methods
- Changed s3 backend's ClearBlock method to not be reliant on knowledge of individual objects
- Added BackendBlock interface and used it in tempodb Find method
This is close to being done, but still requires some additional testing.

**Which issue(s) this PR fixes**:
Fixes #

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by mdisibio — COMMENTED** [review]
Everything is looking good, will do some local testing too.

**Comment by mdisibio** [comment]
Performed full end-to-end testing with all components, looks good.

**Comment by joe-elliott** [comment]
Excellent.  Thanks!  Dropping draft status.  @annanay25 please review when you get a chance.

File: tempodb/backend/s3/compactor.go [hunk 1]
```diff
@@ -48,20 +47,20 @@ func (rw *readerWriter) ClearBlock(blockID uuid.UUID, tenantID string) error {
...
 
-	// list of objects that need to be deleted
-	var delObjects []string
-	delObjects = append(delObjects, util.CompactedMetaFileName(blockID, tenantID))
-	for i := 0; i < bloom.GetShardNum(); i++ {
-		delObjects = append(delObjects, util.BloomFileName(blockID, tenantID, i))
+	// ListObjects(bucket, prefix, marker, delimiter string, maxKeys int)
+	res, err := rw.core.ListObjects(rw.cfg.Bucket, path, "", "/", 0)
+	if err != nil {
+		return errors.Wrapf(err, "error listing tenants in bucket %s", rw.cfg.Bucket)
```
Comment (annanay25) [inline tempodb/backend/s3/compactor.go:56]: nit(non-blocking): This would be `error listing objects in bucket`

File: tempodb/tempodb.go [hunk 2]
```diff
@@ -254,70 +235,20 @@ func (rw *readerWriter) Find(ctx context.Context, tenantID string, id encoding.I
...
-		if err != nil {
-			return nil, fmt.Errorf("error retrieving bloom %v", err)
-		}
-
-		filter := &willf_bloom.BloomFilter{}
-		_, err = filter.ReadFrom(bytes.NewReader(bloomBytes))
-		if err != nil {
-			return nil, fmt.Errorf("error parsing bloom %v", err)
-		}
+		block := encoding.NewBackendBlock(meta)
```
Comment (annanay25) [inline tempodb/tempodb.go:238]: Fantastic!

**Review by annanay25 — APPROVED** [review]
This looks great, awesome work @joe-elliott!

**Pull request merged**